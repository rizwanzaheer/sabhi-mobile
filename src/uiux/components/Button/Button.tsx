import * as React from 'react'
import { TouchableHighlight, TouchableOpacity, ViewStyle } from 'react-native'
import { withTheme } from '../../theming'
import * as Kancha from '../../types'
import Text, { TextTypes } from '../Text/Text'
import Container from '../Container/Container'

export const ButtonBlocks: Kancha.BlocksStatic = {
  Outlined: 'outlined',
  Filled: 'filled',
  Clear: 'clear',
}

interface ButtonProps {
  /**
   * The button type. This sets the theme color
   */
  type?: Kancha.BrandPropOptions
  /**
   * The block appearance of the button
   */
  block?: Kancha.BlockPropsOptions
  /**
   * The text to be displayed
   */
  buttonText?: string

  /**
   * The text to be displayed
   */
  bold?: boolean

  /**
   * Enable smaller button
   */
  small?: boolean

  /**
   * Remove width limitations
   */
  fullWidth?: boolean
  /**
   * The button type. This sets the theme color
   */
  onPress: () => void

  /**
   * Center the button horizontally on screen
   */
  centered?: boolean

  /**
   * Disbable the button
   */
  disabled?: boolean

  /**
   * The button is a navigation button
   */
  navButton?: boolean

  /**
   * Provide an icon component
   */
  icon?: React.ReactNode

  /**
   * Remove button padding for icon buttons
   */
  noPadding?: boolean

  /**
   * Text decoration
   */
  textDecorationLine?: 'none' | 'underline' | 'line-through' | 'underline line-through' | undefined

  /**
   * Provide a testID for e2e tests
   */
  testID?: string

  /**
   * Shadow
   */
  shadowOpacity?: number

  /**
   * Icon button flag
   */
  iconButton?: boolean

  /**
   * Theme used by theme provider
   */
  theme: any
}

const Button: React.FC<ButtonProps> & {
  Block: Kancha.BlocksStatic
} = ({
  type,
  block,
  fullWidth,
  onPress,
  disabled,
  buttonText,
  small,
  centered,
  bold,
  navButton,
  icon,
  noPadding,
  textDecorationLine,
  testID,
  iconButton,
  shadowOpacity,
  theme,
}) => {
  const style: ViewStyle = {
    ...(block && block === 'filled'
      ? { backgroundColor: type ? theme.colors[type].button : theme.colors.primary.button }
      : {}),
    ...(block && block === 'outlined'
      ? {
          backgroundColor: theme.colors.primary.background,
          borderWidth: 2,
          borderColor: type ? theme.colors[type].button : theme.colors.primary.button,
        }
      : {}),
    ...(noPadding || iconButton
      ? {}
      : { padding: small ? theme.spacing.default / 2 : theme.spacing.default }),
    alignItems: 'center',
    ...(fullWidth ? {} : { maxWidth: 300 }),
    borderRadius: theme.roundedCorners.buttons,
    ...(centered ? { alignSelf: 'center' } : {}),
    ...(disabled ? { opacity: 0.2 } : {}),
    ...(shadowOpacity
      ? { shadowColor: '#000000', shadowRadius: 15, shadowOpacity, elevation: shadowOpacity * 10 }
      : {}),
  }

  return navButton ? (
    <TouchableOpacity
      style={style}
      onPress={onPress}
      disabled={disabled}
      testID={testID}
      accessibilityLabel={testID}
    >
      <Text
        textDecorationLine={textDecorationLine}
        type={TextTypes.NavButton}
        buttonTextColor={disabled ? 'secondary' : type}
        block={block}
      >
        {buttonText}
      </Text>
    </TouchableOpacity>
  ) : (
    <TouchableHighlight
      testID={testID}
      accessibilityLabel={testID}
      disabled={disabled}
      onPress={onPress}
      style={style}
      underlayColor={
        block === ButtonBlocks.Clear || iconButton ? 'transparent' : type && theme.colors[type].underlay
      }
    >
      <Container flexDirection={'row'} justifyContent={'center'}>
        {icon && <Container marginRight={10}>{icon}</Container>}
        <Container alignItems={'center'} justifyContent={'center'}>
          <Text
            textDecorationLine={textDecorationLine}
            type={small ? TextTypes.ButtonSmall : TextTypes.Button}
            buttonTextColor={disabled ? 'secondary' : type}
            block={block}
            bold={bold}
          >
            {buttonText}
          </Text>
        </Container>
      </Container>
    </TouchableHighlight>
  )
}

Button.Block = ButtonBlocks

export default withTheme(Button)
