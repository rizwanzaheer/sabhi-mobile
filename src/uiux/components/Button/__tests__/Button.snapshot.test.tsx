import React from 'react'
import { render } from 'react-native-testing-library'

import Button from '../Button'

describe('Component(snapshots): Button', () => {
  it('renders a filled button correctly with all props', () => {
    const onPress = jest.fn()
    const tree = render(
      <Button
        type={'primary'}
        block={'filled'}
        bold={true}
        fullWidth={true}
        centered={true}
        onPress={onPress}
        icon={<React.Fragment />}
        noPadding={true}
        textDecorationLine={'none'}
        testID={'TEST_ID'}
        shadowOpacity={0.5}
      />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders an outlined button correctly', () => {
    const onPress = jest.fn()
    const tree = render(<Button onPress={onPress} block={'outlined'} icon={<React.Fragment />} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders small button correctly', () => {
    const onPress = jest.fn()
    const tree = render(<Button onPress={onPress} small={true} icon={<React.Fragment />} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders an outlined button correctly with a type', () => {
    const onPress = jest.fn()
    const tree = render(
      <Button type={'primary'} onPress={onPress} block={'outlined'} icon={<React.Fragment />} />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders an icon button correctly', () => {
    const onPress = jest.fn()
    const tree = render(<Button onPress={onPress} icon={<React.Fragment />} iconButton={true} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a nav button correctly', () => {
    const onPress = jest.fn()
    const tree = render(<Button onPress={onPress} navButton={true} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a disabled button correctly', () => {
    const onPress = jest.fn()
    const tree = render(<Button onPress={onPress} disabled={true} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a disabled nav button correctly', () => {
    const onPress = jest.fn()
    const tree = render(<Button onPress={onPress} disabled={true} navButton={true} />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
