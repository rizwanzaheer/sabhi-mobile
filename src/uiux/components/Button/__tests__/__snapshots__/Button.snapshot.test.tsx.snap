// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(snapshots): Button renders a disabled button correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 8,
      "maxWidth": 300,
      "opacity": 0.2,
      "padding": 15,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": "center",
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": "center",
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "flexWrap": "wrap",
            "fontSize": 18,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`Component(snapshots): Button renders a disabled nav button correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 8,
      "maxWidth": 300,
      "opacity": 0.2,
      "padding": 15,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#FFFFFF",
        "flexWrap": "wrap",
        "fontSize": 20,
      }
    }
  />
</View>
`;

exports[`Component(snapshots): Button renders a filled button correctly with all props 1`] = `
<View
  accessibilityLabel="TEST_ID"
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#0078ff",
      "borderRadius": 8,
      "elevation": 5,
      "shadowColor": "#000000",
      "shadowOpacity": 0.5,
      "shadowRadius": 15,
    }
  }
  testID="TEST_ID"
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": "center",
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": 10,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    />
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": "center",
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "flexWrap": "wrap",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`Component(snapshots): Button renders a nav button correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 8,
      "maxWidth": 300,
      "opacity": 1,
      "padding": 15,
    }
  }
>
  <Text
    style={
      Object {
        "flexWrap": "wrap",
        "fontSize": 20,
      }
    }
  />
</View>
`;

exports[`Component(snapshots): Button renders an icon button correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 8,
      "maxWidth": 300,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": "center",
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": 10,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    />
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": "center",
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <Text
        style={
          Object {
            "flexWrap": "wrap",
            "fontSize": 18,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`Component(snapshots): Button renders an outlined button correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "#0078ff",
      "borderRadius": 8,
      "borderWidth": 2,
      "maxWidth": 300,
      "padding": 15,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": "center",
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": 10,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    />
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": "center",
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <Text
        style={
          Object {
            "flexWrap": "wrap",
            "fontSize": 18,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`Component(snapshots): Button renders an outlined button correctly with a type 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "#0078ff",
      "borderRadius": 8,
      "borderWidth": 2,
      "maxWidth": 300,
      "padding": 15,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": "center",
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": 10,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    />
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": "center",
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#0078ff",
            "flexWrap": "wrap",
            "fontSize": 18,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`Component(snapshots): Button renders small button correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 8,
      "maxWidth": 300,
      "padding": 7.5,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": "center",
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": 10,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    />
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": undefined,
          "height": undefined,
          "justifyContent": "center",
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <Text
        style={
          Object {
            "flexWrap": "wrap",
            "fontSize": 18,
          }
        }
      />
    </View>
  </View>
</View>
`;
